### ISSUE TITLE
TLS support?

### ISSUE THREAD

**Issue by UladzimirTrehubenka** [panjf2000/gnet#16]
Is TLS support planned?

In accordance with https://github.com/tidwall/evio/issues/28 there are no easy way to implement it:
> The problem is in standard Go TLS implementation. It doesn't have an event-based idea mechanism and actually it can block calls like Reader.Read() during handshake, what is actually not accurate by the io.Reader contract.

Looks like custom TLS implementation is required for this.

**Comment by panjf2000** [comment]
@UladzimirTrehubenka 
It's no plan for now, but I will keep investigating it and support TLS if it is possible.

**Comment by jingc1413** [comment]
hi, brother, When can the tls/dtls be supported?
wait your messge

**Comment by ivanjaros** [comment]
yeah, without tls this is quite useless these days where even services in a private network use encryption. tls is simply a must, no matter which protocol you're using these days, there's no way around it.

**Comment by chowyu08** [comment]
does now support tls?

**Comment by chowyu08** [comment]
什么时候支持tls

**Comment by lesismal** [comment]
I write [nbio](https://github.com/lesismal/nbio), it's another high-performance, event-driven, easy-to-use tcp framework.

I spend some time coping std/crypto/tls, then hack the code to support non-blocking tls for my 

here is the echo example: 
- [echo-server](https://github.com/lesismal/nbio/tree/master/examples/tls/server)
- [echo-client](https://github.com/lesismal/nbio/tree/master/examples/tls/client)

I also write a simple tunnel between the c/s, the tunnel transfers data byte-by-byte to help test tcp-sticky-packet, for more details, see the code here:
- [tcp-sticky-server](https://github.com/lesismal/nbio/tree/master/examples/tls_sticky_packet/server)
- [tcp-sticky--client](https://github.com/lesismal/nbio/tree/master/examples/tls_sticky_packet/client)

the extension only supports the tls1.3 version on the server-side by now, it seems to work well with [nbio](https://github.com/lesismal/nbio).

I may spend more time supporting the versions below tls1.3, and the client-side.

this hack version of tls can be used with other frameworks but need a net.Conn implementation which is used as the under-layer of tls.Conn, it may help, so I opened this https://github.com/panjf2000/gnet/issues/184.

**Comment by lesismal** [comment]
> I write [nbio](https://github.com/lesismal/nbio), it's another high-performance, event-driven, easy-to-use tcp framework.
> 
> I spend some time coping std/crypto/tls, then rewrite it to support non-blocking tls for [nbio](https://github.com/lesismal/nbio).
> 
> here is the echo example:
> 
> * [tls-echo-server](https://github.com/lesismal/nbio/tree/master/examples/tls/server)
> * [tls-echo-client](https://github.com/lesismal/nbio/tree/master/examples/tls/client)
> 
> I also write a simple tunnel between the c/s, the tunnel transfers data byte-by-byte to help test tcp-sticky-packet, for more details, read the code here:
> 
> * [tls-tcp-sticky-server](https://github.com/lesismal/nbio/tree/master/examples/tls_sticky_packet/server)
> * [tls-tcp-sticky-client](https://github.com/lesismal/nbio/tree/master/examples/tls_sticky_packet/client)
> 
> the extension only supports the tls1.3 version, and only on the server-side by now, it seems to work well with [nbio](https://github.com/lesismal/nbio).
> 
> I may spend more time supporting the versions below tls1.3, and the client-side.
> 
> this rewritten version of tls can be used with other frameworks but need a net.Conn implementation which is used as the under-layer of tls.Conn
> 
> I opened this #184, wish my tls could help.

TLS 1.0-1.3 already supported

**Comment by lesismal** [comment]
@UladzimirTrehubenka @jingc1413 @ivanjaros @chowyu08 

how about trying this:
- [tls-echo-server](https://github.com/lesismal/nbio/tree/master/examples/tls/server)
- [tls-echo-client](https://github.com/lesismal/nbio/tree/master/examples/tls/client)

**Comment by lesismal** [comment]
> I write [nbio](https://github.com/lesismal/nbio), it's another high-performance, event-driven, easy-to-use tcp framework.
> 
> I spend some time coping std/crypto/tls, then rewrite it to support non-blocking tls for [nbio](https://github.com/lesismal/nbio).
> 
> here is the echo example:
> 
> * [tls-echo-server](https://github.com/lesismal/nbio/tree/master/examples/tls/server)
> * [tls-echo-client](https://github.com/lesismal/nbio/tree/master/examples/tls/client)
> 
> I also write a simple tunnel between the c/s, the tunnel transfers data byte-by-byte to help test tcp-sticky-packet, for more details, read the code here:
> 
> * [tls-tcp-sticky-server](https://github.com/lesismal/nbio/tree/master/examples/tls_sticky_packet/server)
> * [tls-tcp-sticky-client](https://github.com/lesismal/nbio/tree/master/examples/tls_sticky_packet/client)
> 
> the extension only supports the tls1.3 version, and only on the server-side by now, it seems to work well with [nbio](https://github.com/lesismal/nbio).
> 
> I may spend more time supporting the versions below tls1.3, and the client-side.
> 
> this rewritten version of tls can be used with other frameworks but need a net.Conn implementation which is used as the under-layer of tls.Conn
> 
> I opened this #184, wish my tls could help.


TLS1.0-1.3 and client-side support has been done

**Comment by panjf2000** [comment]
之前的消息就算了，以后不要再在我的项目 issue 里宣传你自己的库了。
@lesismal

**Comment by lesismal** [comment]
ok

**Comment by HeisenbergV** [comment]
请问tls还会支持吗

**Comment by panjf2000** [comment]
Sorry, I don't have a particularly convincing idea of how to implement TLS in `gnet` at the present stage, and I don't have much time to do this at the moment, it'd be great if anyone could take this job and open a PR for it, contribute code or provide some insightful thoughts.

**Comment by vikassandhu999** [comment]
> Sorry, I don't have a particularly convincing idea of how to implement TLS in `gnet` at the present stage, and I don't have much time to do this at the moment, it'd be great if anyone could take this job and open a PR for it, contribute code or provide some insightful thoughts.

Though I don't have much experience with TLS implementations but If someone is willing to take the charge of the job, I can assist them.

**Comment by seymourtang** [comment]
Here is an example on how to implement TLS in gnet。[https://github.com/luyu6056/tls](url)

**Comment by rock59** [comment]
s2n-tls is a small, modern, secure and fast TLS implementation developed and used by AWS. It provides posix-like api so it seems easy to integrate it in gent for developers who dont mind using cgo.

**Comment by ruimya** [comment]
@panjf2000  Has gnet considered which version will officially support TLS? Additionally, is there a way to minimize intrusion into TLS source code and avoid maintaining another set of TLS source code? Can we simply call non blocking code from the crypto/tls and rewrite the blocking code in gnet? just immature suggestion

**Comment by panjf2000** [comment]
Sorry, I can't give you a specific ETA for `gnet` TLS. It's not a trivial job, I don't have much time for that, also I don't have the urge to use TLS on `gnet` cuz the major companies that are currently using `gnet` employ TLS in API gateway, thus `gnet` itself doesn't need it. If somebody wants to implement TLS for `gnet`, I'd love to review it.
